ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -229,52 +230,9 @@ public boolean updateAccounts(Collection<Account> accounts) {
       logger.debug("Accounts={} conflict with the accounts in local cache. Cancel the update operation.", accounts);
       accountServiceMetrics.updateAccountErrorCount.inc();
     } else {
-      hasSucceeded = helixStore.update(FULL_ACCOUNT_METADATA_PATH, currentData -> {
-        ZNRecord newRecord;
-        if (currentData == null) {
-          logger.debug(
-              "ZNRecord does not exist on path={} in HelixPropertyStore when updating accounts. Creating a new ZNRecord.",
-              FULL_ACCOUNT_METADATA_PATH);
-          newRecord = new ZNRecord(ZN_RECORD_ID);
-        } else {
-          newRecord = currentData;
-        }
-        Map<String, String> accountMap = newRecord.getMapField(ACCOUNT_METADATA_MAP_KEY);
-        if (accountMap == null) {
-          logger.debug("AccountMap does not exist in ZNRecord when updating accounts. Creating a new accountMap");
-          accountMap = new HashMap<>();
-        }
-        AccountInfoMap remoteAccountInfoMap;
-        try {
-          remoteAccountInfoMap = new AccountInfoMap(accountMap);
-        } catch (JSONException e) {
-          // Do not depend on Helix to log, so log the error message here.
-          logger.error("Exception occurred when building AccountInfoMap from accountMap={}", accountMap, e);
-          accountServiceMetrics.remoteDataCorruptionErrorCount.inc();
-          throw new IllegalStateException("Exception occurred when building AccountInfoMap from accountMap", e);
-        }
-        // if there is any conflict with the existing record, fail the update. Exception thrown in this updater will
-        // be caught by Helix and helixStore#update will return false.
-        if (hasConflictingAccount(accounts, remoteAccountInfoMap)) {
-          // Throw exception, so that helixStore can capture and terminate the update operation
-          throw new IllegalArgumentException(
-              "Updating accounts failed because one account to update conflicts with existing accounts");
-        } else {
-          for (Account account : accounts) {
-            try {
-              accountMap.put(String.valueOf(account.getId()), account.toJson().toString());
-            } catch (Exception e) {
-              String message = "Updating accounts failed because unexpected exception occurred when updating accountId="
-                  + account.getId() + " accountName=" + account.getName();
-              // Do not depend on Helix to log, so log the error message here.
-              logger.error(message, e);
-              throw new IllegalStateException(message, e);
-            }
-          }
-          newRecord.setMapField(ACCOUNT_METADATA_MAP_KEY, accountMap);
-          return newRecord;
-        }
-      }, AccessOption.PERSISTENT);
+      ZkUpdater zkUpdater = new ZkUpdater(accounts);
+      hasSucceeded = helixStore.update(FULL_ACCOUNT_METADATA_PATH, zkUpdater, AccessOption.PERSISTENT);
+      zkUpdater.persistBackup(hasSucceeded);
```

Comment:
i understand what you are trying to do here - removing the disk write op out of the `update()` call but I'd like to understand why that is so important. If update is a rare operation and there are no timeouts set in helix for how long it should take, then we should be able to write the file(s) out in the `update()` method itself.

Coming to why that is useful, imagine that it so happens that this persist call fails while constructing the "new state" (or for any other reason other than the "previous state" itself failing - for e.g while writing to the file) and the new state is bad but has been updated in helix, then we would not have a backup to fall back on. I'd much rather that we persist the old state **_as soon as_** we get it from zk before doing anything else. And then persist the new state **_as soon as_** it is written into zk. I know this will create two files (as opposed to the one you have right now) but I think it provides something more reliable than doing one persist after zk has been updated - don't want to be left in a situation without any backup and the state is bad.

Reply:
If the zk level version does not match and update is called again, it perhaps means that persistence again is a desired property - is that what will happen?

As for the second issue you bring up (not knowing if the update succeeded or not), what do you think about writing the "new state" outside of the `update()` function ? (i.e. where you do it currently). Then we can combine the writing of the new state and the success status into one file. I guess the spirit of this change is that the "previous state" is more important than the "new state" - the new one is only for informational purposes.